lib/components/user/favorite-location.js

```diff
@@ -0,0 +1,236 @@
+ */
+function makeFavoriteLocation (baseLocation, type, icon) {
+  const { lat, lon, name } = baseLocation
+  return {
+    address: name,
+    icon,
+    lat,
+    lon,
+    name,
+    type
+  }
+}
+
+/**
+ * Renders a user favorite location,
+ * and lets the user edit the details (address, type, TODO: nickname) of the location.
+ */
+class FavoriteLocation extends Component {
+  constructor () {
+    super()
+    this.state = {
+      newLocation: null
+    }
+  }
+
+  _handleDelete = () => {
+    const { arrayHelpers, index, isFixed, location } = this.props
+    if (isFixed) {
+      // For 'Home' and 'Work', replace with the default blank locations instead of deleting.
+      let newLocation
+      if (isHome(location)) {
+        newLocation = BLANK_HOME
+      } else if (isWork(location)) {
+        newLocation = BLANK_WORK
+      }
+
+      this._handleLocationChange({ location: newLocation })
+    } else {
+      arrayHelpers.remove(index)
+    }
+  }
+
+  _handleLocationChange = ({ location }) => {
+    const { arrayHelpers, index, location: oldLocation } = this.props
+    const { icon, type } = oldLocation
+    arrayHelpers.replace(index, makeFavoriteLocation(location, type, icon))
+  }
+
+  _handleLocationTypeChange = ({ icon, type }) => {
+    const { arrayHelpers, index, location } = this.props
+    arrayHelpers.replace(index, makeFavoriteLocation(location, type, icon))
+  }
+
+  _handleNew = ({ location }) => {
+    const { arrayHelpers } = this.props
+
+    // Set, then unset the location state to reset the LocationField after address is entered.
+    // (both this.setState calls should trigger componentDidUpdate on LocationField).
+    this.setState({ newLocation: location })
+
+    arrayHelpers.push(makeFavoriteLocation(location, 'custom', 'map-marker'))
+
+    this.setState({ newLocation: null })
+  }
+
+  render () {
+    const { index, isFixed, isNew } = this.props
+
+    // When true, isNew makes the control add a new location instead of editing an existing one.
+    // The appearance is also different to differentiate the control with others.
+    let icon
+    let iconTitle
+    let handler
+    // Don't display a delete button if the item is fixed and its address is blank,
+    // or for the new location row.
+    let hideDelete
+    let location
+    let placeholder
+    let DropdownBtn
+    if (isNew) {
+      location = this.state.newLocation
+      icon = 'plus'
+      iconTitle = null
+      handler = this._handleNew
+      hideDelete = true
+      placeholder = 'Add another place'
+      DropdownBtn = NewLocationDropdownButton
+    } else {
+      location = this.props.location
+      // TODO: lookup icon using location.type
+      icon = location.icon
+      iconTitle = location.type
+      handler = this._handleLocationChange
+      hideDelete = false // isFixed && isBlank(location.address)
+      placeholder = isFixed ? `Add ${location.type}` : 'Enter a favorite address'
+      DropdownBtn = StyledDropdown
```

Comment:
Since so much is different between new/existing, it might be more readable to have a separate component: 
```jsx
<NewLocation arrayHelpers={arrayHelpers} />
```

Reply:
Updated in https://github.com/opentripplanner/otp-react-redux/pull/280/commits/9298f94739c2196c0a6f47717a9481c12ad1e306